C++: Port PrivateData.qll from C# and use it in cpp/cleartext-transmission

geoffw0 opened the pull request:
Port `PrivateData.qll` from C#.
 - the C# version was chosen as a base, rather than the version in experimental as had been originally planned.  They appear to be extremely similar, but the C# version is more complete.
 - the experimental version has been deprecated.
 - additional enhancements / tuning has been done to maximize results with minimal FPs.

Use `PrivateData.qll` in `cpp/cleartext-transmission`.
 - on LGTM we find a small number of new results, generally I think they're good, a few are weak or FP results.
 - `cpp/cleartext-storage-buffer`, `cpp/cleartext-storage-file` and `cpp/cleartext-storage-database` might also benefit, but I need to further investigate whether results are good for these queries.

MathiasVP commented on cpp/ql/lib/semmle/code/cpp/security/PrivateData.qll:
```diff
@@ -0,0 +1,63 @@
+      // Government identifiers, such as Social Security Numbers
+      "%social%security%",
+      // Contact information, such as home addresses and telephone numbers
+      "%postcode%", "%zipcode%", "%telephone%",
```
Since we already allow arbitrary strings between the word `bank` and `account`, how about we also allow arbitrary strings between `post` and `code`, and `zip` and `code`?

Our tests would improve if we added `%licese%key%` as well (or maybe just `%license%`?). Do we catch more good results if we add those?

(These issues aren't at all blocking. Feel free to postpone them to another PR if you agree that this is something we should do)

geoffw0 replied:
> Since we already allow arbitrary strings between the word bank and account, how about we also allow arbitrary strings between post and code, and zip and code?

I did some test runs on LGTM and this turned out to be problematic.  `post`, `code` and `zip` are commonly used in variable and function names referring to technical things, so we get a few too many false positives.

I may try switching from `match` to regexps, and allow exactly 0 or 1 arbitrary character in between words in all of these cases - rather than doing the best we can with the options of any string (often too permissive) or nothing (too restrictive).

> Our tests would improve if we added %licese%key% as well (or maybe just %license%?). Do we catch more good results if we add those?

Planned as a follow-up, as I think that is closer to a password than private personal data so should be a change in `SensitiveExprs.qll` rather than `PrivateData.qll`.

geoffw0 replied:
I've now addressed your first point, the second will be addressed in a follow-up PR.

geoffw0 replied:
Follow-up PR here: https://github.com/github/codeql/pull/8623

geoffw0 commented:
I've just pushed some changes:
 - updated the approach from using `match` to `regexpMatch`, which allows slightly more accurate expressions and thus slightly more accurate results (at least on tests).  A repeat LGTM run appears to have identical results.  Local testing suggests performance hasn't been affected much.
 - I've added a couple more test cases around the ones for `salary`, I wasn't happy with them as they weren't really exercising the existing code properly.
 - I've added the [CWE-359](https://cwe.mitre.org/data/definitions/359.html) tag to this query, as it now covers many of the types of information described in that CWE.

geoffw0 commented:
... and I've deleted the experimental `PrivateData.qll` now as well.

MathiasVP reviewed (approved):
LGTM!

geoffw0 commented:
Ah, thanks for doing a DCA run.  I was distracted yesterday and didn't get around to it.
 - there's a lost query result; it isn't from a query affected by `PrivateData.qll` or related changes here, so I think its wobble.
 - there are new extraction errors in `microsoft__calculator`; this PR contains no extractor changes and this project commonly wobbles, so again I think no cause for concern.
 - overall analysis time is almost unchanged.

geoffw0 commented:
Merging...

MathiasVP commented:
> Ah, thanks for doing a DCA run. I was distracted yesterday and didn't get around to it.

No worries. It was really for testing something completely different 😂. But I'm glad you found it useful!
